### PR TITLE
explain how to create a machine

### LINKED ISSUE #25: Provide extensive documentation on how to implement a state machine
Explain the data types involves in the definition, how they relate among each other, and how to concretely implement a machine

### PR DESCRIPTION
**Pull request by marcosh**
I'm pointing this PR not to `main` because it refers to concepts which are not in `main` yet.

The idea of what I wrote so far is to explain all the bits and pieces which one needs to know to create a `StateMachine a b`, including the meaning, the type and the kind of all the parameters used around.

So far, I just did the easy bit. The hard part is to describe how to use properly the `BaseMachine` type. In the `spec/CRM/Example` folder there are several examples which hopefully illustrate how to use such a type. More examples, more complex and illustrative ones, are coming soon.

closes #25

### CODE REVIEW

File: docs/how-to-create-a-machine.md
Comment (AlexeiDrake) [inline docs/how-to-create-a-machine.md:49]: I certainly appreciate the effort that must have gone into drawing this intricate ASCII art. :laughing: 

Consider using [Mermaid](https://mermaid.js.org/syntax/stateDiagram.html) instead.

```mermaid
stateDiagram-v2
    direction LR
    [*] --> A
    state A {
        direction LR
        [*] --> B: a
        B --> C: b
        C --> [*]: c
    }
    A --> [*]
```

**Review by AlexeiDrake — CHANGES_REQUESTED** [review]
I'll hold off on reviewing this until the previous PRs are merged.

File: docs/how-to-create-a-machine.md
Comment (marcosh) [reply docs/how-to-create-a-machine.md:49]: I do love Mermaid, but for these diagram I kind of prefer to be able to control precisely how things are represented, especially for more complex diagrams like this one:

Screenshot_20230131_095000

If in the future we will use Mermaid to generate similar diagrams automatically, then it would definitely make sense to use the same tool in the documentation

**Comment by AlexeiDrake** [comment]
I'm really surprised at how big this PR has become. Why are there new decisions and code changes in the diff?

**Comment by marcosh** [comment]
> I'm really surprised at how big this PR has become. Why are there new decisions and code changes in the diff?
> 
> Can we split this up into multiple PRs that all target `main`?

I rebased this PR on the of the latest developments, so that I could document those, too. It would probably be easier to review commit by commit. Otherwise, if you prefer, I could split the PR in several ones which will need to be reviewed one after the other

File: spec/CRM/Example/RiskManager/Projection.hs
Comment (AlexeiDrake) [inline spec/CRM/Example/RiskManager/Projection.hs:32]:
```suggestion
r1 <> r2 = ReceivedData
    { receivedUserData = receivedUserData r1 <|> receivedUserData r2
    , receivedLoanDetails = receivedLoanDetails r1 <|> receivedLoanDetails r2
    , receivedCreditBureauData = receivedCreditBureauData r1 <|> receivedCreditBureauData r2
    }
```

File: spec/CRM/RiskManagerSpec.hs
Comment (AlexeiDrake) [inline spec/CRM/RiskManagerSpec.hs:44]:
```suggestion
  CreditBureauData
    { missedPaymentDeadlines = MissedPaymentDeadlines 2
    , arrears = EuroCents 100000
    }
```

File: docs/how-to-create-a-machine.md
Comment (AlexeiDrake) [inline docs/how-to-create-a-machine.md:24]:
```suggestion
Every state machine that is not `Basic` is a composition of two state machines, which are ultimately expressed in terms of `Basic`.

Let's skip over `Basic` for now, and look at the other constructors first.
```

This is why I think `Compose` should be named `Sequence`; all of these constructors are compositions, but not every composition is sequential.

File: docs/how-to-create-a-machine.md
Comment (AlexeiDrake) [inline docs/how-to-create-a-machine.md:154]: Case in point that `Compose` is a misleading constructor name. :)

File: docs/how-to-create-a-machine.md
Comment (AlexeiDrake) [inline docs/how-to-create-a-machine.md:202]:
```suggestion
We use the `[b]` notation to denote multiple outputs, but we can in fact use any `Foldable` type.
```

File: docs/how-to-create-a-machine.md
Comment (AlexeiDrake) [inline docs/how-to-create-a-machine.md:247]:
```suggestion
A straightforward approach is to create a sum type of vertices:
```

File: src/CRM/StateMachine.hs
Comment (AlexeiDrake) [reply src/CRM/StateMachine.hs]: I see this was done in a later commit. :)

File: src/CRM/Graph.hs
Comment (marcosh) [reply src/CRM/Graph.hs:44]: definitely more readable, thanks!

File: src/CRM/BaseMachine.hs
Comment (marcosh) [reply src/CRM/BaseMachine.hs:163]: this is actually not used, so I'll just remove it.

Anyway, I call it `sequence` because what it does is commuting two effects, like `sequence` does, but the type signature is not the same and actually not compatible

Comment (marcosh) [reply src/CRM/BaseMachine.hs:163]: maybe this could actually be useful, so I'll keep it instead...

It could be generalised to

```haskell
sequence
  :: Functor f
  => ActionResult Identity topology state initialVertex (f output)
  -> ActionResult f topology state initialVertex output
```

On the other hand, `sequenceA` from `Traversable` would need to be

```haskell
sequence
  :: Functor f
  => ActionResult m topology state initialVertex (f output)
  -> f (ActionResult m topology state initialVertex output)
```

I'll add a comment to explain this

File: src/CRM/BaseMachine.hs
Comment (marcosh) [reply src/CRM/BaseMachine.hs:142]: while with `fmap` you change with a function the type of your last parameter `output` in this case, with `hoist` you use a natural transformation (i.e. a parametric function, that's why the `forall` needs to be explicit) to change the context: in this case `m` becomes `n`.

It's the same idea you use in `Servant` to change context, see https://docs.servant.dev/en/stable/tutorial/Server.html#using-another-monad-for-your-handlers

File: spec/CRM/Example/RiskManager/Projection.hs
Comment (marcosh) [reply spec/CRM/Example/RiskManager/Projection.hs:32]: makes sense, but it still needs the `Last` and `getLast`

File: spec/CRM/RiskManagerSpec.hs
Comment (marcosh) [reply spec/CRM/RiskManagerSpec.hs:44]: that's better, thanks

File: docs/how-to-create-a-machine.md
Comment (marcosh) [reply docs/how-to-create-a-machine.md:24]: yes, now I agree with you. I'll do the change in a separate PR

File: docs/how-to-create-a-machine.md
Comment (marcosh) [reply docs/how-to-create-a-machine.md:247]: I also thought initially to say `sum type`, but also `Either Int String` is a sum type and that probably won't work here because then it won't satisfy the `Demote vertex ~ vertex` constraint we have in the `Basic` constructor for `StateMachine`.

In fact, I need to add some more information about those constraints and what they imply (basically they imply that the type must be finite)

**Pull request closed without merging**
